### ISSUE TITLE
ML Request: Predict an optimal stopping point, for longer test runtimes

### ISSUE THREAD

**Issue by oindrillac** [aicoe-aiops/ocp-ci-analysis#226]
Based on previous test and build time duration data, we can predict and suggest a stopping point, beyond which it is likely to result in a failure.

**Comment by aakankshaduggal** [comment]
Hema and I are looking into this issue and we have a few questions and suggestions here. For predicting an optimal stopping point beyond which it is likely to result in a failure, we think we can start by looking into these metrics :

* **probability to fail**: At a given moment in time, what is the probability that your test will fail.
* **time unit failure**:  That is, how many days, weeks, months, hours, minutes or seconds do you have until the test in question stops working.

We were hoping to use regression to calculate the above-mentioned metrics. 

Do you think this is a good way to start?
@MichaelClifford 
cc : @hemajv 

[Reference](https://medium.com/swlh/converting-a-probability-to-fail-into-a-time-to-failure-metric-9a0ddf6122c4)

**Comment by MichaelClifford** [comment]
@hemajv @aakankshaduggal :+1: 

 > probability to fail

How do you plan on calculating this value? I don't think this is a metric that we currently track. There are probably some more robust, feature intensive way to do this, but for an initial attempt I was thinking that for a test (or set of tests) you could look at the distribution of test duration for all failed tests and the distribution of test duration for all the successful tests (ideally generating two distinct distributions). Then given a new duration value, evaluate the likelihood that it belongs to one of the two distributions. I think this will give you a way to calculate the "probability to fail" based solely on the duration metric.   

One thing to note with the proposed method above is that we are building the distributions based on complete test times, but the usefulness of the tool would be to evaluate time taken so far in the middle of a run. So we may have to add some manual threshold value to recommended a reasonable stopping point: For example, when there is a 75% likelihood the current duration is from the "fail" distribution stop the test.    

> time unit failure

If we have the above probability to fail, I'm not sure we need to project a time to failure as well. But let me know if I misunderstood something : ) 

@oindrillac is this in-line with your original idea for this ML Request?

**Comment by hemajv** [comment]
> > probability to fail
> 
> How do you plan on calculating this value? I don't think this is a metric that we currently track. There are probably some more robust, feature intensive way to do this, but for an initial attempt I was thinking that for a test (or set of tests) you could look at the distribution of test duration for all failed tests and the distribution of test duration for all the successful tests (ideally generating two distinct distributions). Then given a new duration value, evaluate the likelihood that it belongs to one of the two distributions. I think this will give you a way to calculate the "probability to fail" based solely on the duration metric.
> 
> One thing to note with the proposed method above is that we are building the distributions based on complete test times, but the usefulness of the tool would be to evaluate time taken so far in the middle of a run. So we may have to add some manual threshold value to recommended a reasonable stopping point: For example, when there is a 75% likelihood the current duration is from the "fail" distribution stop the test.
> 
Yes, this is something that we don't have currently. We were hoping to make use of some of the metrics from the [persistent_failures](https://github.com/aicoe-aiops/ocp-ci-analysis/blob/master/notebooks/data-sources/TestGrid/metrics/persistent_failures_analysis.ipynb) notebook and as you mentioned look at the test duration distributions to calculate the probability.

> > time unit failure
> 
> If we have the above probability to fail, I'm not sure we need to project a time to failure as well. But let me know if I misunderstood something : )
> 
So I think here ideally we wanted to forecast and identify the time range during which test failures are likely to occur in the future. We thought this can be a next step for helping answer additional questions such as predicting the root cause of the test failures.

**Comment by aakankshaduggal** [comment]
> How do you plan on calculating this value? I don't think this is a metric that we currently track. There are probably some more robust, feature intensive way to do this, but for an initial attempt I was thinking that for a test (or set of tests) you could look at the distribution of test duration for all failed tests and the distribution of test duration for all the successful tests (ideally generating two distinct distributions). Then given a new duration value, evaluate the likelihood that it belongs to one of the two distributions. I think this will give you a way to calculate the "probability to fail" based solely on the duration metric.

This seems like a good way to start. Thanks for your input.
> If we have the above probability to fail, I'm not sure we need to project a time to failure as well. But let me know if I misunderstood something : )

For the above, I mean if we have the probability to fail we can predict the optimal stopping point using regression as Hema mentioned too. We plan to build a model that uses the probability to fail to predict the time to failure.

**Comment by MichaelClifford** [comment]
:+1: Ok, It sounds like there are a couple of ideas that can be explored to meet this ML Requests. That said, I think its important that you nail down a single ML task to peruse first (we can always extend it later). So it sound like the two options right now are 1) Using some set of features about a test (that need to be determined), train a regression model to predict the possible time to failure. or 2) Empirically determine the distribution of failed and passed test durations and use hypothesis testing to determine the likelihood a test will fail given its current duration, or 3) :man_shrugging: some new idea.

You may want to break this down into a couple of smaller issues where you collect and evaluate the data and features you intended to use, define a performance criteria you want to evaluate your work with and prototype a model, etc.  

In either case, let's try and define the ML task to pursue in this issue as specifically and machine learn-y as possible :smiley:

**Comment by aakankshaduggal** [comment]
Sure @MichaelClifford. Thank you for your suggestions. Definitely going to create a smaller issues and get started 😄

**Comment by oindrillac** [comment]
@hemajv @aakankshaduggal can we close this given we have a service now?
